conmon/conmon.c

```diff
@@ -1321,6 +1321,18 @@ int main(int argc, char *argv[])
 		if (dup2(slavefd_stderr, STDERR_FILENO) < 0)
 			pexit("Failed to dup over stderr");
 
+		/* If LISTEN_PID env is set, we need to modify to set
+		   it to the new child process */
+		char *listenpid = getenv("LISTEN_PID");
+		if (listenpid != NULL) {
```

Comment:
Instead of clobbering `LISTEN_PID` in all cases, you'll want to only update here if the *previous* `LISTEN_PID` matched the initial `conmon` PID.  The full chain would be:

1. Caller delegates to `conmon`.
2. `conmon` delegates to runtime's `create`.
3. Runtime's `create` delegates to the container process.

If the caller delegated to a PID that didn't match `conmon`, I think we have to assume they knew what they were doing and leave `LISTEN_PID` alone.

Reply:
I understand the reason to do this, but it also makes the parent much more difficult to do.  Which puts a lot of burden on the user.  Having a go program without fork and exec setup something like this ends up very complicated like in runc.  I would like to avoid this complexity for very little benefit.

Reply:
> I understand the reason to do this, but it also makes the parent much more difficult to do. Which puts a lot of burden on the user.   Having a go program without fork and exec setup something like this ends up very complicated like in runc.

I'm not saying “drop `conmon`” or anything drastic like that.  I agree that `conmon`'s current fork and exec approach is easier to do in C than it would be in Go.  I just don't want to clobber `LISTEN_PID` if the caller intended to delegate to some other process.  The change vs. your current ac1e320310 would be changing this conditional to:

```c
if (listenpid != NULL) {
    long int listenpid_int = strtol(listenpid, NUL, 10);
    if (long_int == conmon_pid) {
        …your current pidstr clobber…
    }
}
```

with (somewhere before [the first `fork`][1]):

```c
pid_t conmon_pid = getpid();
```

I don't see how that puts an increased burden on the user or how it would make the parent more difficult to use.  In most cases, I expect `LISTEN_PID` to be set by systemd like it already does (e.g. [here][2]).  Do you see users frequently setting it by hand?  Or are you saying that because systemd is launching `crio` or similar, that the PID value wouldn't match `conmon`'s initial PID, and that to get that match you'd need:

1. Caller delegates to `crio`.
2. `crio` delegates to `conmon`.
3. `conmon` delegates to runtime's `create`.
4. Runtime's `create` delegates to the container process.

and that working up the `crio` delegation would be too difficult?
[1]: https://github.com/kubernetes-incubator/cri-o/blob/5345c6299aac5a0899947eebe82e3dca8d4ad5d9/conmon/conmon.c#L1174
[2]: https://github.com/systemd/systemd/blob/v237/src/activate/activate.c#L226

Reply:
Our use case is

systemd -> podman (GO Based Docker CLI clone) -> conmon -> OCIRUntime -> pid1
The issue is the complextty of getting podman to do a fork/set environment variable/exec conmon.

One potential compromize would be to add a --clober-listenfd flag to ignore your check to conmon, and then we can just have podman call conmon with that option.

Reply:
> One potential compromize would be to add a `--clobber-listenfd` flag to ignore your check to conmon, and then we can just have podman call conmon with that option.

Yeah, that works for me.  It's unfortunate that container creation chains have so many Go → C-shim transitions, and having a `--clobber-listenfd` boolean seems like a lesser evil than having podman add another C shim.